week-1-ts-ecosystem/w01e03-object-oriented-structure/src/shop/cart.class.ts

```diff
@@ -0,0 +1,72 @@
+import {FreeProductInterface, ProductInterface} from "./product.interface";
+import {CartType} from "./cart.type";
+
+
+export default class Cart {
+    #products: (ProductInterface | FreeProductInterface)[] = [];
+    #type: CartType;
+    #sum = 0;
+
+    constructor(type: CartType) {
+        this.#type = type;
+    }
+
+    addItem(product: ProductInterface | FreeProductInterface): void {
+        const existingProduct = this.#products.findIndex(item => item.id === product.id);
+
+        if (existingProduct > -1) {
+            this.updateCart(product.id);
+            return;
+        }
+
+        if (product.type === this.#type) {
+            this.#products.push({...product, amount: 1});
+        }
+
+        if (product.type !== 'FREE') {
+            this.#sum += product.price;
+        }
+    }
+
+    updateCart(id: string) {
+        this.#products = this.#products.map(item => {
```

Comment:
Nie musisz tworzyć a każdym updatem nowej tablicy ;) 
Wystarczy skorzystać z referencji do obektu znalezionego np. przez `this.getItem(id)`

Zobacz również że Twoje `updateCart` robi coś konkretniejszego niż `updateCart`
robi `increaseProductAmount`

Bo nie aktualizuje produktu tylko zwiększa jego ilość jeźeli istnieje.